internal/policy/container/has_modified_files.go

```diff
@@ -24,140 +28,151 @@ var _ check.Check = &HasModifiedFilesCheck{}
-	pkgList, err := rpm.GetPackageList(ctx, imgRef.ImageFSPath)
+	layerDir, err := afero.TempDir(fs, "", "rpm-layers-")
 	if err != nil {
-		return nil, fmt.Errorf("could not get rpm list: %w", err)
+		return nil, nil, err
 	}
+	defer func() {
+		_ = fs.RemoveAll(layerDir)
+	}()
 
-	packageFiles, err := p.getInstalledFilesFor(pkgList)
-	if err != nil {
-		return nil, fmt.Errorf("could not list installed files: %w", err)
+	if imgRef.ImageInfo == nil {
+		return nil, nil, fmt.Errorf("image reference invalid")
 	}
 
 	layers, err := imgRef.ImageInfo.Layers()
 	if err != nil {
-		return nil, fmt.Errorf("could not get image layers: %w", err)
+		return nil, nil, err
 	}
 
-	files := make([][]string, 0, len(layers))
+	layerIDs := make([]string, 0, len(layers))
+	layerRefs := make(map[string]packageFilesRef, len(layers))
 
-	// Uncompress each layer and build a slice containing the files
-	// modified by each layer.
-	for _, layer := range layers {
-		r, err := layer.Uncompressed()
+	// Uncompress each layer and build maps containing the packages,
+	// the package files, and the files modified by each layer
+	// Also generate a list of the layerIDs so we can keep the
+	// order within the maps.
+	for idx, layer := range layers {
+		layerIDHash, err := layer.DiffID()
 		if err != nil {
-			return nil, fmt.Errorf("could not extract layers: %w", err)
-		}
-		pathChan := make(chan string)
-
-		var wg sync.WaitGroup
-		wg.Add(1)
-		go func() {
-			// For each path in the pathChan, add it to the layer's
-			// list of files.
-			layerFiles := make([]string, 0)
-			for path := range pathChan {
-				layerFiles = append(layerFiles, path)
-			}
-			// also add it to the overall list of files.
-			files = append(files, layerFiles)
-			wg.Done()
-		}()
-		// add paths to the pathChan
-		err = untar(pathChan, r)
+			return nil, nil, fmt.Errorf("unable to retrieve diff id for layer: %w", err)
+		}
+		layerID := layerIDHash.String()
+
+		layerDir := filepath.Join(layerDir, layerID)
+		err = fs.Mkdir(layerDir, 0o755)
 		if err != nil {
-			return nil, fmt.Errorf("failed to extract tarball: %w", err)
+			return nil, nil, fmt.Errorf("could not create layer directory: %w", err)
 		}
 
-		wg.Wait() // wait for file list to get appended to the files var
-	}
+		layerIDs = append(layerIDs, layerID)
 
-	files, dropped := p.dropFirstLayerIfEmpty(files)
-	if dropped {
-		// tell the user the first layer was dropped
-		diff0, _ := layers[0].DiffID()
-		diff1, _ := layers[1].DiffID()
-		logger.V(log.DBG).Info(
-			fmt.Sprintf("The first layer (%s) contained no files, so the next layer (%s) is being used as the base layer.",
-				diff0.String(),
-				diff1.String(),
-			))
-	}
+		files, err := generateChangesFor(layer)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	return &packageFilesRef{files, packageFiles}, nil
-}
+		found, pkgList := findRPMDB(ctx, layer)
+		if !found {
+			logger.V(log.TRC).Info("could not find rpm database")
```

Comment:
This runs on every layer right? I think adding layer context might be useful here.

```suggestion
			logger.V(log.TRC).Info("could not find rpm database in layer", "layer", layerID)
```

That code may not be correct. Making the assumption that this accepts keys/values. Either way, you get the idea.

Reply:
Sure. I'll add that context.